distributed/core.py

```diff
@@ -594,29 +592,33 @@ async def handle_stream(self, comm, extra=None, every_cycle=[]):
             await comm.close()
             assert comm.closed()
 
-    @gen.coroutine
-    def close(self):
+    async def close(self):
         for pc in self.periodic_callbacks.values():
             pc.stop()
         for listener in self.listeners:
             future = listener.stop()
             if inspect.isawaitable(future):
-                yield future
+                await future
         for i in range(20):  # let comms close naturally for a second
             if not self._comms:
                 break
             else:
-                yield asyncio.sleep(0.05)
-        yield [comm.close() for comm in list(self._comms)]  # then forcefully close
+                await asyncio.sleep(0.05)
+        await asyncio.gather(
+            *[comm.close() for comm in list(self._comms)]
+        )  # then forcefully close
         for cb in self._ongoing_coroutines:
             cb.cancel()
         for i in range(10):
             if all(c.cancelled() for c in self._ongoing_coroutines):
                 break
             else:
-                yield asyncio.sleep(0.01)
+                await asyncio.sleep(0.01)
 
         self._event_finished.set()
+        # Allow event listeners a cycle to react before we return
+        # See also https://github.com/dask/distributed/pull/4734
+        await asyncio.sleep(0)
```

Comment:
I received errors with still pending tasks from the worker after shutdown

```
  <Task pending name='Task-41' coro=<Worker.ensure_computing() running at
```

letting the loop run for another cycle here seemed to do the trick but frankly, I don't understand why. Especially not what this has to do with my proposed changes. Is anybody able to shed light on this?

Reply:
I think that `await asyncio.sleep(0)` may not actually let the event loop cycle.  I think that `sleep(0)` might special-case and continue directly with the same coroutine.  You might want to try something like `sleep(1e-6)` to see if that makes a difference.

Reply:
It is a special case but it does allow one cycle, see https://github.com/python/cpython/blob/accea7dc2bd30a6e8e1b0334acfca9585cbd7f8a/Lib/asyncio/tasks.py#L585-L601 

If this doesn't work I'll try a fixed sleep but this feels fragile

Do you have an idea how this could be connected?

Reply:
I'm not sure.

Maybe there was a final task submission that previously was ignored but is now actually submitted?